### PR TITLE
Gh 2211 functions

### PR DESCRIPTION
**Pull request by p013570**

### CODE REVIEW

File: core/data/src/main/java/uk/gov/gchq/gaffer/data/element/function/ReduceRelatedElements.java
Comment (GCHQDev404) [inline core/data/src/main/java/uk/gov/gchq/gaffer/data/element/function/ReduceRelatedElements.java]: * Copyright 2020 Crown Copyright

Comment (GCHQDev404) [inline core/data/src/main/java/uk/gov/gchq/gaffer/data/element/function/ReduceRelatedElements.java]: 2020

File: core/data/src/main/java/uk/gov/gchq/gaffer/data/element/function/ReduceRelatedElements.java
Comment (GCHQDev404) [inline core/data/src/main/java/uk/gov/gchq/gaffer/data/element/function/ReduceRelatedElements.java:2]: is 2017 correct ?

File: core/data/src/main/java/uk/gov/gchq/gaffer/data/element/function/TypeSubTypeValueToTuple.java
Comment (GCHQDev404) [inline core/data/src/main/java/uk/gov/gchq/gaffer/data/element/function/TypeSubTypeValueToTuple.java:2]: If these are all new please examine all the copyright dates.

**Review by GCHQDev404 — CHANGES_REQUESTED** [review]

**Comment by n3101** [comment]
@GCHQDev404 I have contacted @p013570 offline. He said: 

"I don't have an environment setup and not even sure if I have my login details for github. As it is such a basic change any other developer could easily do that."

I will assign this to one of our team to make the changes.

**Review by p3430233 — APPROVED** [review]

File: core/data/src/main/java/uk/gov/gchq/gaffer/data/element/function/ReduceRelatedElements.java
Comment (GCHQDev404) [inline core/data/src/main/java/uk/gov/gchq/gaffer/data/element/function/ReduceRelatedElements.java:2]: new Class with copyright starting from 2017? This occurs many times.

File: core/data/src/test/java/uk/gov/gchq/gaffer/data/element/function/ReduceRelatedElementsTest.java
Comment (GCHQDev404) [inline core/data/src/test/java/uk/gov/gchq/gaffer/data/element/function/ReduceRelatedElementsTest.java:42]: I'm seeing warning about incomplete abstract methods that have not been implemented.
Please review, and amend. I'm seeing this many test locations.

**Review by GCHQDev404 — CHANGES_REQUESTED** [review]

**Review by p3430233 — COMMENTED** [review]
TypeSubTypeValueToTupleTest.java, method getExpectedSignatureOutputClasses() passes  when return type is set to String.class but was expecting tuple as output.

File: core/data/src/main/java/uk/gov/gchq/gaffer/data/element/function/ReduceRelatedElements.java
Comment (GCHQDev404) [inline core/data/src/main/java/uk/gov/gchq/gaffer/data/element/function/ReduceRelatedElements.java:49]: This is out of date due to how old this PR is.
New classes need this value to be updated for the version it will be added to.
1.16.2 I think will be the next version. 
This is a number that needs to be chased until it gets merged in.
please review for all other New classes.

Comment (n3101) [reply core/data/src/main/java/uk/gov/gchq/gaffer/data/element/function/ReduceRelatedElements.java:49]: Next version will be 1.17

**Review by GCHQDev404 — CHANGES_REQUESTED** [review]

**Comment by GCHQDev404** [comment]
@p3430233 

Something odd is happening if I add this it doesn't error
```
   @Override
    protected Class[] getExpectedSignatureOutputClasses() {
        throw new RuntimeException();
//        return new Class[]{Tuple.class};
    }
```

**Comment by GCHQDev404** [comment]
blocked by gh-2443

**Comment by n3101** [comment]
Blocked label removed as #2443 has been resolved and closed

File: core/data/src/main/java/uk/gov/gchq/gaffer/data/element/function/ReduceRelatedElements.java
Comment (t92549) [reply core/data/src/main/java/uk/gov/gchq/gaffer/data/element/function/ReduceRelatedElements.java]: Next version is now 1.18.0

**Comment by t92549** [comment]
The test function `shouldGenerateExpectedOutputSignature` didn't exist in Koryphe when gh-2239 was first submitted, so this only became an issue recently, as the Koryphe dependency was updated. 

Tests were failing when return type of `TypeValueToTuple.apply` was `Tuple<String>`. A `TypeValueTuple` is returned, but this implements `Tuple<String>` so this should be okay:
https://github.com/gchq/Gaffer/blob/cdcf9045f110435462fbb2cbb7a0b1c813fabfff/core/data/src/main/java/uk/gov/gchq/gaffer/data/element/function/TypeValueToTuple.java#L29-L36

When a Tuple was expected as output, the tests failed:
https://github.com/gchq/Gaffer/blob/ce58f484de859d1a271ced1cb5c31735cf2a05e0/core/data/src/test/java/uk/gov/gchq/gaffer/data/element/function/TypeValueToTupleTest.java#L111-L112

However, having looked at Koryphe, the `Signature` made from a Tuple is a `TupleSignature`.
https://github.com/gchq/koryphe/blob/5389edb0415650ddb2d05a9468c8f6e55443e2ca/core/src/main/java/uk/gov/gchq/koryphe/signature/Signature.java#L173.

When a `TupleSignature` is checked if `assignable` from another type, it is not testing if the Tuple class can be assigned from the other type. Instead, it gets the Tuple types (ie. String in `Tuple<String>`), and checks if they are assignable.
https://github.com/gchq/koryphe/blob/5389edb0415650ddb2d05a9468c8f6e55443e2ca/core/src/main/java/uk/gov/gchq/koryphe/signature/TupleSignature.java#L72

That is why when the test was changed to expect `String.class` it passed. This implies strange behaviour in `TupleSignature` in Koryphe, so more investigation will have to be done there.

In the meantime though I changed the return type in the class and method to `TypeValueTuple`:
https://github.com/gchq/Gaffer/blob/b7c76eb042624e478d233b8c06ffd9141d0018ad/core/data/src/main/java/uk/gov/gchq/gaffer/data/element/function/TypeValueToTuple.java#L28-L30

Not only is this cleaner but it actually makes the tests pass as `TupleSignature` is bypassed. I think this solution is okay but we will have to look into why `TupleSignature` works as it does (checking the type within the Tuple) and whether it is actually used anywhere.

File: core/data/src/main/java/uk/gov/gchq/gaffer/data/element/function/TypeSubTypeValueTuple.java
Comment (t511203) [inline core/data/src/main/java/uk/gov/gchq/gaffer/data/element/function/TypeSubTypeValueTuple.java:25]: Formatting tidy up here.

File: core/data/src/main/java/uk/gov/gchq/gaffer/data/element/function/TypeSubTypeValueTuple.java
Comment (t511203) [inline core/data/src/main/java/uk/gov/gchq/gaffer/data/element/function/TypeSubTypeValueTuple.java:28]: Missing Javadoc.

File: core/data/src/main/java/uk/gov/gchq/gaffer/data/element/function/TypeSubTypeValueTuple.java
Comment (t511203) [inline core/data/src/main/java/uk/gov/gchq/gaffer/data/element/function/TypeSubTypeValueTuple.java:32]: Might be safer to remove this telescoping constructor behaviour.

File: core/data/src/main/java/uk/gov/gchq/gaffer/data/element/function/TypeValueTuple.java
Comment (t511203) [inline core/data/src/main/java/uk/gov/gchq/gaffer/data/element/function/TypeValueTuple.java:32]: Same comment here regarding telescoping constructors.

**Review by t511203 — COMMENTED** [review]
I'm happy to make these changes, then for someone else to sign this off and hit merge.

File: core/data/src/main/java/uk/gov/gchq/gaffer/data/element/function/TypeSubTypeValueTuple.java
Comment (t92549) [reply core/data/src/main/java/uk/gov/gchq/gaffer/data/element/function/TypeSubTypeValueTuple.java:28]: Also worth updating these as the next release this will go into will be 1.19.0 now.

File: core/data/src/main/java/uk/gov/gchq/gaffer/data/element/function/TypeSubTypeValueTuple.java
Comment (GCHQDev404) [reply core/data/src/main/java/uk/gov/gchq/gaffer/data/element/function/TypeSubTypeValueTuple.java]: Why? the issue with telescoping contractors behaviour is the parameter list grows and grows, this has 1 thing in it.

File: core/data/src/test/java/uk/gov/gchq/gaffer/data/element/function/TypeValueToTupleTest.java
Comment (t92549) [inline core/data/src/test/java/uk/gov/gchq/gaffer/data/element/function/TypeValueToTupleTest.java:112]: Perhaps this should be changed back to `Class[]{String.class}`. Despite the slightly confusing layout, checking that the `Tuple` contains the correct type is potentially more useful, as explained in https://github.com/gchq/koryphe/issues/225#issuecomment-896004455.

File: core/data/src/main/java/uk/gov/gchq/gaffer/data/element/function/TypeValueToTuple.java
Comment (t92549) [inline core/data/src/main/java/uk/gov/gchq/gaffer/data/element/function/TypeValueToTuple.java:30]: Perhaps this should be changed back to `Tuple<String>`. This ensures that the output signature can be properly tested, as explained below. As well as this, it is more obvious what should go inside the `Tuple`. If this is the case, then perhaps this should also be changed in `TypeSubTypeValueToTuple` as well as some other places in the code base too?

Comment (t92549) [reply core/data/src/main/java/uk/gov/gchq/gaffer/data/element/function/TypeValueToTuple.java:30]: Or perhaps creating and testing for a specific `TypeValueTuple` is preferred?

File: core/data/src/test/java/uk/gov/gchq/gaffer/data/element/function/TypeValueToTupleTest.java
Comment (t92549) [reply core/data/src/test/java/uk/gov/gchq/gaffer/data/element/function/TypeValueToTupleTest.java:112]: After some more understanding, it is okay for Gaffer to implement its own concrete implementations of a `Tuple` and test for those types (see https://github.com/gchq/koryphe/issues/225#issuecomment-896103601). No changes should be required here.

**Review by t92549 — APPROVED** [review]

**Review by p3430233 — APPROVED** [review]

**Comment by n3101** [comment]
OK, I'm really confused about the state of this PR! It has two approvals from @p013570 and @t92549 but still awaiting a third review from @GCHQDev404! Is this really intended or can this be merged?

**Comment by t92549** [comment]
> OK, I'm really confused about the state of this PR! It has two approvals from @p013570 and @t92549 but still awaiting a third review from @GCHQDev404! Is this really intended or can this be merged?

I think it is pretty much done but it was me and @p013570 that did a lot of the work so it would perhaps be better if someone else had a final look over and review

File: core/data/src/main/java/uk/gov/gchq/gaffer/data/element/function/TypeSubTypeValueTuple.java
Comment (t11947) [inline core/data/src/main/java/uk/gov/gchq/gaffer/data/element/function/TypeSubTypeValueTuple.java:50]: Might be worth sticking these literals in constants considering they're reused a bit

File: core/data/src/test/java/uk/gov/gchq/gaffer/data/element/function/ReduceRelatedElementsTest.java
Comment (t11947) [inline core/data/src/test/java/uk/gov/gchq/gaffer/data/element/function/ReduceRelatedElementsTest.java:261]: I'd probably find another way to get the test to pass

File: core/data/src/test/java/uk/gov/gchq/gaffer/data/element/function/ReduceRelatedElementsTest.java
Comment (t11947) [inline core/data/src/test/java/uk/gov/gchq/gaffer/data/element/function/ReduceRelatedElementsTest.java:385]: Consider using assertThrows

File: core/data/src/test/java/uk/gov/gchq/gaffer/data/element/function/ReduceRelatedElementsTest.java
Comment (t11947) [inline core/data/src/test/java/uk/gov/gchq/gaffer/data/element/function/ReduceRelatedElementsTest.java:266]: Might be worth making these constants at the class level if they're reused in multiple tests

File: core/data/src/test/java/uk/gov/gchq/gaffer/data/element/function/ReduceRelatedElementsTest.java
Comment (t11947) [inline core/data/src/test/java/uk/gov/gchq/gaffer/data/element/function/ReduceRelatedElementsTest.java:394]: It might be worth including some more granular testing of the (currently private) logic found in some of the helper methods.

File: core/operation/src/main/java/uk/gov/gchq/gaffer/operation/function/ToTrailingWildcardPair.java
Comment (t11947) [inline core/operation/src/main/java/uk/gov/gchq/gaffer/operation/function/ToTrailingWildcardPair.java:54]: Is it worth adding null checking/defaulting for if endOfRange is null?

**Review by t11947 — CHANGES_REQUESTED** [review]
Just a couple of small bits then it should be good to go. Previous changes appear to have been actioned.

File: core/data/src/test/java/uk/gov/gchq/gaffer/data/element/function/ReduceRelatedElementsTest.java
Comment (t11947) [reply core/data/src/test/java/uk/gov/gchq/gaffer/data/element/function/ReduceRelatedElementsTest.java:394]: Might be worth extracting the updater method to its own class/set of helper methods to keep encapsulation.

**Review by t11947 — APPROVED** [review]

**Pull request merged**
